experimental/dds/tree2/src/core/tree/detachedFieldIndex.ts

```diff
@@ -141,29 +157,105 @@ export class DetachedFieldIndex {
 	}
 
 	public encode(): string {
-		const data: [Major, Minor, ForestRootId][] = [];
-		forEachInNestedMap(this.detachedNodeToField, (root, key1, key2) => {
-			data.push([key1, key2, root]);
-		});
-		return JSON.stringify({
-			data,
-			id: this.rootIdAllocator.getNextId(),
-		});
+		return this.codec.encode({
+			data: this.detachedNodeToField,
+			maxId: this.rootIdAllocator.getNextId(),
+		}) as string;
 	}
 
 	/**
 	 * Loads the tree index from the given string, this overrides any existing data.
 	 */
 	public loadData(data: string): void {
-		const detachedFieldIndex: { data: readonly [Major, Minor, ForestRootId][]; id: number } =
-			JSON.parse(data);
-		const map = new Map();
-		for (const [major, minor, root] of detachedFieldIndex.data) {
-			setInNestedMap(map, major, minor, root);
-		}
-		this.detachedNodeToField = map;
+		const detachedFieldIndex: {
+			data: NestedMap<Major, Minor, ForestRootId>;
+			maxId: number;
+		} = this.codec.decode(data);
+
 		this.rootIdAllocator = idAllocatorFromMaxId(
-			detachedFieldIndex.id,
+			detachedFieldIndex.maxId,
 		) as IdAllocator<ForestRootId>;
+		this.detachedNodeToField = detachedFieldIndex.data;
 	}
 }
+
+// Define the Major and Minor types:
+const MajorSchema = Type.Union([Type.String(), Type.Number(), Type.Undefined()]);
+const MinorSchema = Type.Number();
+
+const ForestRootIdSchema = brandedNumberType<ForestRootId>();
+
+// Define the tuple:
+const TupleSchema = Type.Tuple([MajorSchema, MinorSchema, ForestRootIdSchema]);
+
+export const Format = Type.Object({
+	version: Type.Literal(version),
+	data: Type.Array(TupleSchema),
+	maxId: ForestRootIdSchema,
+});
+
+export type Format = Static<typeof Format>;
+
+const Versioned = Type.Object({
+	version: Type.Number(),
+});
+type Versioned = Static<typeof Versioned>;
+
+export function makeDetachedNodeToFieldCodec({
+	jsonValidator: validator,
+}: ICodecOptions): IJsonCodec<{
+	data: NestedMap<Major, Minor, ForestRootId>;
+	maxId: ForestRootId;
+}> {
```

Comment:
```suggestion
}, string> {
```

this should let you avoid casting above.

Reply:
updated codec type and removed unnecessary casting